File: ruby/beer-song/beer_song.rb
class BeerSong
  def verse(number)
    case number
    when 0
      zero_verse
    when 1
      one_verse
    when 2
      two_verse
    else
      common_verse(number)
    end
  end

  def verses(from, to)
    (to..from).to_a.reverse.reduce('') do |result, number|
      result + verse(number) + "\n"
    end
  end

  def sing
    verses(99, 0)
  end

  private

  def common_verse(number)
    "#{number} bottles of beer on the wall, #{number} bottles of beer.\n"\
    "Take one down and pass it around, #{number-1} bottles of beer on the wall.\n"
  end

  def one_verse
    "1 bottle of beer on the wall, 1 bottle of beer.\n" \
    "Take it down and pass it around, no more bottles of beer on the wall.\n"
  end

  def two_verse
    "2 bottles of beer on the wall, 2 bottles of beer.\n" \
    "Take one down and pass it around, 1 bottle of beer on the wall.\n"
  end

  def zero_verse
    "No more bottles of beer on the wall, no more bottles of beer.\n" \
    "Go to the store and buy some more, 99 bottles of beer on the wall.\n"
  end
end

File: ecmascript/beer-song/beer-song.js
export default class Beer {
  static verse(number) {
    return this.firstLine(number) + this.secondLine(number);
  }

  static sing(start = 99, end = 0) {
    let result = [];
    for(let number = start; number >= end; number -= 1) {
      result.push(this.firstLine(number) + this.secondLine(number));
    }
    return result.join("\n");
  }

  static firstLine(number) {
    let bottle = this.pluralizeBottle(number);
    number = number === 0 ? "no more" : number;
    return `${this.capitalize(number)} ${bottle} of beer on the wall, ${number} ${bottle} of beer.\n`;
  }

  static secondLine(number) {
    if(number === 0) {
      return "Go to the store and buy some more, 99 bottles of beer on the wall.\n";
    }

    let newNumber = number - 1;
    if(newNumber === 0) { newNumber = "no more"; }

    let downCount = number === 1 ? "it" : "one";
    let bottle = this.pluralizeBottle(newNumber);
    return `Take ${downCount} down and pass it around, ${newNumber} ${bottle} of beer on the wall.\n`;
  }

  static capitalize(string) {
    string = (string).toString();
    return string.slice(0, 1).toUpperCase() + string.slice(1);
  }

  static pluralizeBottle(number) {
    return number === 1 ? "bottle" : "bottles";
  }
}
